### ISSUE TITLE
DwC Importer - Integrate identifiedById and recordedById fields

### ISSUE THREAD

**Issue by mjy** [SpeciesFileGroup/taxonworks#2025]
These fields are identical to their non "Id" fields, but include globally unique identifiers instead.  

Rules for integration:
* A: If they are provided, then they must either 1) be accompanied by their corresponding non "id" field OR 2) match an existing identifier in TW.  If not 1 or 2 then Error is added to the row.
* B: If no error in A, then find Person based on global ID, if found, _add name string components as AlternateValues if they differ_.  If no Person is found create the person, and add the global ID.

**Comment by LocoDelAssembly** [comment]
Are there existing samples to aid development of this feature?

**Comment by debpaul** [comment]
Hi Hernán,

I might be able to get you some from GBIF. I think? that the GBIF Norway is now sharing these data for many of their collectors...

Would that be helpful?

Debbie
________________________________

**Comment by mjy** [comment]
@dshorthouse @debpaul somethign we can get from Bionomia directly?

**Comment by dshorthouse** [comment]
Yes, but not as DwC-A or identifiedByID/recordedByID am afraid. These would be in frictionless data packages, eg https://bionomia.net/dataset/68513375-3aa5-4f6f-9975-d97d56c21d61. Whatever route you take, a sub-ticket here is to resolve the identifier before incorporating it. The identifiers in recordedByID or identifiedByID are just that – a delimited list. When I scour these from GBIF every two weeks, there are often ...umm... problems. Resolving them can be done via wikidata but it's an additional step that will considerably decelerate your DwC importer routine.

**Comment by LocoDelAssembly** [comment]
@debpaul if you get it with little effort yes please. If none of the future targets have this then it will become handy to use in the meantime.

Thanks @dshorthouse will take a look.

**Comment by debpaul** [comment]
To @LocoDelAssembly 
 - [one GBIF dataset](https://www.gbif.org/dataset/26f5b360-8770-4d54-9c2d-397798a5e513) I found (of several) has 408,129 records --> so I made you [a subset csv file of these records with 171,729 rows](https://www.dropbox.com/s/mhrskvqjr0o16dw/norway_IDonly_csvformat.csv?dl=0) where either or both of these field sets of interest are populated.

- dwc:recordedBy and dwc:recordedByID
- dwc:identifiedBy and dwc:identifiedByID

In this dataset, 

1. I believe, we see max two collectors listed in each / any (although other datasets could have more).
2. NOTE: some records have no dwc:recordedByID or identifiedByID (I've removed record that were completely blank for both of these).
3. And there is another case that is NOT in this dataset. You could have **2** or more name strings in dwc:recordedBy, but in dwc:recordedByID perhaps only one of these people has a known Identifier. So you'd have 
   - | | where there was a value and then no value for the other person. The "pipe" might help figure out who the identifier refers to (or it, the pipe, might be absent in this case).
4. This dataset DOES have a person where there are 2 identifiers for this person in both dwc:identifiedByID and dwc:recordedByID.
   - example: the ORCID and Wikidata QID here are BOTH for Leif Aarvik with no identifier for Nini.
   - this is why (at least one reason why) @dshorthouse hints that one needs to resolve these identifiers first.

`Leif  Aarvik, Nini C. Aarvik | https://orcid.org/0000-0002-0112-8837|https://www.wikidata.org/wiki/Q17114254`

Original Dataset is: [Entomology, Natural History Museum, University of Oslo](https://www.gbif.org/dataset/26f5b360-8770-4d54-9c2d-397798a5e513)

There are [more datasets from the collections in Norway uploaded to GBIF](https://www.gbif.org/dataset/search?type=OCCURRENCE&publishing_country=NO), that have data in these fields. Just let me know if you need more examples after this. @dshorthouse has Loads of experience with these now.

**Comment by LocoDelAssembly** [comment]
Thanks much @debpaul! I think one is OK now.

**Comment by dshorthouse** [comment]
This might be useful: https://github.com/bionomia/bionomia/blob/master/lib/identifier.rb

**Comment by mjy** [comment]
Thanks @dshorthouse.  We have a library of validators similar for all our Identifier classes.  In other words we should extract from both and create a gem sometime.

**Comment by dshorthouse** [comment]
@mjy Yeah, figured you would. Feel free to grab that ORCID checksum check in the meantime. That's a good one to verify a validly-formed ORCID w/o having to bang on an API.

**Comment by mjy** [comment]
Aha!  Nice, this we don't have, thanks.

https://github.com/bionomia/bionomia/blob/master/lib/identifier.rb#L61

**Comment by LocoDelAssembly** [comment]
https://github.com/tdwg/dwc/issues/102

Will have to make an exception for the importer to accept these and re-upload the sample file. It is in http://rs.gbif.org/core/dwc_occurrence_2020-07-15.xml (with GBIF's own namespace) but not in http://rs.gbif.org/core/dwc_occurrence.xml

**Comment by mjy** [comment]
@LocoDelAssembly what is the role of extensions?  If the data come in as extension, or the namespace references an extension (real or contrived), would it be easier to import these data.  I.e. set ourselves up for generically handing/importing extension related data.

**Comment by debpaul** [comment]
@mjy @LocoDelAssembly I can imagine (and @dshorthouse might comment) the Agents Extension playing a role in the nearish future. (being developed at TDWG, lead by @dshorthouse)

**Comment by dshorthouse** [comment]
fwiw, the AgentActions DwC extension (in pre-draft form) will not have the terms `identifiedByID` or `recordedByID` because the agent's action (many possible values) is split off from the agent's identifier. When this goes to production, we'll all need significant, custom code to handle all the variability.

**Comment by tmcelrath** [comment]
Not currently implemented in DWCA Importer. Reopen? Would be a nice feature.

**Comment by debpaul** [comment]
@pnhogan requesting we reopen this issue so that we can
- import recordedByID and identifiedByID successfully
- and helps us to reduce duplication of people that need unifying.
